### PR TITLE
Upgrade D3 and d3-flameGraph

### PR DESCRIPTION
**Pull request by Krinkle**
This builds on top of https://github.com/perftools/xhgui/pull/215.

The upgrade is a fairly large commit because D3 and d3-flameGraph have to be updated at the same time, and they bring with them a few breaking changes that required the frontend glue in `flamegraph.twig` and `xhgui-charts.js` to be changed a little bit.

### CODE REVIEW

**Comment by Krinkle** [comment]
When testing on this branch I notice that these upstream improvements expose a pre-existing bug in XHGui.

First, the upstream improvements:

1. Chart labels now use their own unique class name (`d3-flame-graph-label`) and no longer clash with Bootstraps' `label` class. This removes various unwarranted style conflicts and fixes the legibility issue raised in https://github.com/perftools/xhgui/issues/211.

2. Previously d3-flameGraph did not use the `value` of the stack data directly, but instead computed its own node values by recursively adding together all child nodes. This caused issues described at https://github.com/perftools/xhgui/issues/212 and https://github.com/perftools/xhgui/issues/207.  In short: All values, except the leaf nodes where way off. Often making the higher level nodes (including `main()`) sometimes report a value 10x or 100x higher than it should be. This has been fixed.

This second point, means the size of nodes in the flame graph is now accurate, and the hover tip shows the correct microsecond values now.

However, this given the width of the rectangles now being computed directly based on the node's own value relative to `main`, this means the values have to be valid. And it turns out, the data exported by XHGui via `/run/flamegraph/data?id=:id` is not always valid. At least, in my own testing I found that whenever the same function is called from two subtrees, the value of the last occurrence is wrongly re-used for all earlier references. Causing weird graphs like this:

```
1. main() - 43,000 µs
1.1. Something\here() - 1,800 µs # does not match 1,300 + 6,500
1.1.1. Something\else() – 1,300 µs
1.1.2. spl_autoload_call – 6,500 µs # was probably 300 µs?
1.2. Other\stuff() – 34,700 µs
1.3. spl_autoload_call – 6,500 µs
```

In current master this data corruption is not obvious at first because flameGraph computes the width based on their children, which means "Something\here" would (wrongly) be given `1800+1300+6500`.

Instead, on this pull requests' branch, the graph can now be rendered in a way that has various rectangles overlapping each other because the children can sometimes be wider than their parent:

| Before (master) | After (this PR)
|--|--
| screen shot 2017-11-13 at 21 55 09 | screen shot 2017-11-13 at 21 55 37


Note in the after image how some of the rectangles now have rectangles transparently behind them.

File: webroot/css/xhgui.css [hunk 1]
```diff
@@ -376,6 +376,64 @@ tr .no-results {
...
+    color: #fff;
+}
+
+.d3-flame-graph-tip {
+  line-height: 1;
+  padding: 12px;
+  background: #fff;
+  border-radius: 6px;
+  box-shadow: 0 5px 10px rgba(0, 0, 0, 0.2);
+  border: 1px solid rgba(0, 0, 0, 0.2);
```
[the remark this answers is not in the document]
Comment (Krinkle) [reply webroot/css/xhgui.css]: (pre-existing issue, unrelated)

File: src/templates/runs/flamegraph.twig [hunk 2]
```diff
@@ -72,12 +70,6 @@ d3.json("{{ url('run.flamegraph.data', {id: profile.id|trim }) }}", function(err
         return console.warn(error);
     }
 
-    flameGraph.height(getDepth(data.data) * cellHeight);
-
-    flameGraph.sort(function (a, b) {
```
Comment (Krinkle) [inline src/templates/runs/flamegraph.twig]: Note to self: This is actually not redundant. Due to the code accessing `data.name` I mistakenly thought this is a name sort, but it's actually looking in XHGui's data object, not the D3 `d.data` object looks similar.

This is from fb8fac599e88d863e33095927a0e6f4d8bb7ff44 and needs to stay.

**Comment by markstory** [comment]
The children being wider than the parents results in some very funny looking graphs locally.  I'll see if I can get the calculations behind the flamegraph to add up better. The tests you contributed in #217 will be helpful in verifying any fixes.

**Comment by Krinkle** [comment]
@markstory I see the tests were merged, but I'm unsure as to whether the problem was solved. Do all nodes now have values that are never less than the total of direct children?

Let me know if/when it is resolved, or what I can do to help, would be nice to move this along.

While the master version (last I checked) didn't appear broken, as mentioned in the previous comments, it is also broken due to the nodes being recursively summed. This has two notable side-effects: Proportions are off (small values are shown too small, large values are shown too large), and the total is way too high. However, it does have the benefit of masking the problem we found in this pull request.

**Comment by markstory** [comment]
I don't think the root problem was solved. Last I remember there were different display issues with these changes. I haven't had much time to dedicate to this project as of late though.

**Comment by Krinkle** [comment]
Closing in favour of #246.

**Pull request closed without merging**